mlir/lib/Transform/EliminateInitLoops.cpp

```diff
@@ -0,0 +1,126 @@
+#include "Transform/TTLPasses.h"
+#include <llvm/Support/Debug.h>
+
+namespace mlir {
+namespace ttl {
+#define GEN_PASS_DEF_TTLELIMINATEINITLOOPS
+#include "Transform/TTLPasses.h.inc"
+
+#define DEBUG_TYPE "ttl-eliminate-init-loops"
+
+namespace {
+
+class EliminateInitLoops
+    : public impl::TTLEliminateInitLoopsBase<EliminateInitLoops> {
+public:
+  using impl::TTLEliminateInitLoopsBase<
+      EliminateInitLoops>::TTLEliminateInitLoopsBase;
+
+  void runOnOperation() final {
+
+    func::FuncOp func = getOperation();
+    LLVM_DEBUG(llvm::dbgs() << "Function name: " << func.getName() << "\n");
+    func.walk([&](ttl::ForLoop forLoop) {
+      if (eliminateLoopIfPossible(forLoop)) {
+        ++numLoopsEliminated;
+      }
+    });
+  }
+
+private:
+  bool eliminateLoopIfPossible(ttl::ForLoop forLoop) {
+    // To eliminate an initialization loop, it must fulfill a number of
+    // conditions:
+
+    // 1. The loop must be directly in the function, i.e., not a nested loop or
+    // inside an `if`.
+    if (!isa<func::FuncOp>(forLoop->getParentOp())) {
+      LLVM_DEBUG(llvm::dbgs() << "Failed condition 1\n");
+      return false;
+    }
+
+    // 2. The loop's step must be a constant 1 and the start and end of the loop
+    // must be constants.
+    auto maybeConstantStep =
+        dyn_cast_or_null<ttl::IntConstant>(forLoop.getStep().getDefiningOp());
+    auto maybeConstantStart = dyn_cast_or_null<ttl::IntConstant>(
+        forLoop.getLowerBound().getDefiningOp());
+    auto maybeConstantEnd = dyn_cast_or_null<ttl::IntConstant>(
+        forLoop.getUpperBound().getDefiningOp());
+    if (!maybeConstantStep || maybeConstantStep.getConstVal() != 1 ||
+        !maybeConstantStart || !maybeConstantEnd) {
+      LLVM_DEBUG(llvm::dbgs() << "Failed condition 2\n");
+      return false;
+    }
+
+    auto tensorSize =
+        maybeConstantEnd.getConstVal() - maybeConstantStart.getConstVal();
+
+    // 3. The loop must return a single result of a 1-dimensional tensor and
+    // take a single init val of tensor type. The tensor must have static size
+    // and must match the range between the constant start and end of the loop.
+    if (forLoop.getNumResults() != 1 || forLoop.getInitArgs().size() != 1) {
+      LLVM_DEBUG(llvm::dbgs() << "Failed condition 3-1\n");
+      return false;
+    }
+    auto resultTy =
+        dyn_cast<ttl::TensorType>(forLoop.getResults().front().getType());
+    auto initArgTy =
+        dyn_cast<ttl::TensorType>(forLoop.getInitArgs().front().getType());
+    if (!resultTy || !initArgTy || resultTy.getShape().size() != 1 ||
+        resultTy != initArgTy || resultTy.getShape().front() != tensorSize) {
+      LLVM_DEBUG(llvm::dbgs() << "Failed condition 3-2\n");
+      return false;
+    }
+
+    // 4. The init val of the loop must be a TensorEmpty which only has a single
+    // user (the loop).
```

Comment:
Nit: This one feels a bit artificial; couldn't you just conditionally erase the tensor init if it has users after replacing the loop?

Reply:
After online discussion, I've realized that the users affected by the pass are only users of the init loop, so other uses of the `TensorEmpty` are save. This has been fixed in the latest commit.